webapi/src/main/scala/org/knora/webapi/messages/admin/responder/projectsmessages/ProjectsMessagesADM.scala

```diff
@@ -505,130 +506,61 @@ case class ProjectADM(
-  ): ProjectIdentifierADM = {
-
-    val parametersCount: Int = List(
-      maybeIri,
-      maybeShortname,
-      maybeShortcode
-    ).flatten.size
-
-    // something needs to be set
-    if (parametersCount == 0) throw BadRequestException("Empty project identifier is not allowed.")
-
-    if (parametersCount > 1) throw BadRequestException("Only one option allowed for project identifier.")
-
-    new ProjectIdentifierADM(
-      maybeIri =
-        sf.validateAndEscapeOptionalProjectIri(maybeIri, throw BadRequestException(s"Invalid user project $maybeIri")),
-      maybeShortname = sf.validateAndEscapeOptionalProjectShortname(
-        maybeShortname,
-        throw BadRequestException(s"Invalid user project shortname $maybeShortname")
-      ),
-      maybeShortcode = sf.validateAndEscapeOptionalProjectShortcode(
-        maybeShortcode,
-        throw BadRequestException(s"Invalid user project shortcode $maybeShortcode")
-      )
-    )
+  case class Iri(value: String) extends ProjectIdentifierADM {
+    def apply(value: String): ProjectIri = ProjectIri.make(value).fold(e => throw e.head, v => v)
   }
-}
 
-/**
- * Represents the project's identifier. It can be an IRI, shortcode or shortname.
- *
- * @param maybeIri       the project's IRI.
- * @param maybeShortname the project's shortname.
- * @param maybeShortcode the project's shortcode
- */
-class ProjectIdentifierADM private (
-  maybeIri: Option[IRI] = None,
-  maybeShortname: Option[String] = None,
-  maybeShortcode: Option[String] = None
-) {
-
-  // squash and return value.
-  val value: String = List(
-    maybeIri,
-    maybeShortname,
-    maybeShortcode
-  ).flatten.head
-
-  def hasType: ProjectIdentifierType =
-    if (maybeIri.isDefined) {
-      ProjectIdentifierType.IRI
-    } else if (maybeShortcode.isDefined) {
-      ProjectIdentifierType.SHORTCODE
-    } else {
-      ProjectIdentifierType.SHORTNAME
-    }
+  case class Shortcode(value: String) extends ProjectIdentifierADM {
+    def apply(value: String): ShortCode = ShortCode.make(value).fold(e => throw e.head, v => v)
+  }
 
-  /**
-   * Tries to return the value as an IRI.
-   */
-  def toIri: IRI =
-    maybeIri.getOrElse(
-      throw DataConversionException(s"Identifier $value is not of the required 'ProjectIdentifierType.IRI' type.")
-    )
+  case class Shortname(value: String) extends ProjectIdentifierADM {
+    def apply(value: String): ShortName = ShortName.make(value).fold(e => throw e.head, v => v)
+  }
 
-  /**
-   * Returns an optional value of the identifier.
-   */
-  def toIriOption: Option[IRI] =
-    maybeIri
+  case class Uuid(value: String) extends ProjectIdentifierADM
 
-  /**
-   * Tries to return the value as an SHORTNAME.
-   */
-  def toShortname: String =
-    maybeShortname.getOrElse(
-      throw DataConversionException(s"Identifier $value is not of the required 'ProjectIdentifierType.SHORTNAME' type.")
-    )
+  def asIriOption(id: ProjectIdentifierADM): Option[String] =
```

Comment:
These `as...Option` methods could become methods in the trait

```scala
sealed trait ProjectIdentifierADM {
  def asIri(): Option[IRI] =
    this match {
      case Iri(value) => Some(value)
      case _          => None
    }
}
```

Reply:
Is there more difference between these 2 ideas instead of less code (no param in methods). We keep methods in companion objects here...